### ISSUE TITLE
Deprecation Warnings For LWRPs

### ISSUE THREAD

**Issue by jaym** [chef/chef#3621]
#3603 introduced more regressions. This code is very complex, and it's purpose is to display a deprecation warning. I think we need to reduce our risk here and simply this down. We have 2 proposals here, both involve using the old behavior unless `Chef::Config[:treat_deprecation_warnings_as_errors]` is `true`.

Proposal one:
-------------------
In this case, we do not set `Chef::Resource::LwrpName` if `:treat_deprecation_warnings_as_errors`. This will fail hard on any access the `Chef::Resource::LwrpName`
```ruby
def self.register_deprecated_lwrp_class(resource_class, class_name)
  if !Chef::Config[:treat_deprecation_warnings_as_errors]
    Chef::Resource.const_set(class_name, resource_class)
  end
end
```

Proposal two:
------------------
We register a class that will fail if initialized when `:treat_deprecation_warnings_as_errors` is set. There are some cases where this does not work, like `my_resource.is_a?(Chef::Resource::LwrpFoo)`
```ruby
class ClassThatFailsIfInitialized
  def initialize
    Chef::Log.deprecation("You cannot access LWRP's through Chef::Resource and Chef::Provider")
  end
end
 
def self.register_deprecated_lwrp_class(resource_class, class_name)
  if Chef::Config[:treat_deprecation_warnings_as_errors]
    Chef::Resource.const_set(class_name, ClassThatFailsIfInitialized)
  else
    Chef::Resource.const_set(class_name, resource_class)
  end
end
```
cc @chef/client-maintainers

**Comment by danielsdeleo** [comment]
Option two seems preferable at first glance. :+1: on abandoning deprecation warnings for this, there's just too many edge cases that are very common and are getting broken.

**Comment by lamont-granquist** [comment]
:+1: on option 2 i guess

if we have to do this it kind of scuttles the idea of doing rolling deprecations without major version bumps.

**Comment by danielsdeleo** [comment]
@lamont-granquist I think there's other ways to go forward, like wrapping all LWRP creation in `Chef.create_provider(api_level = 1) do; end`, the old stuff becomes version 0 and we can detect very easily what version you're on.

**Comment by jkeiser** [comment]
:+1: on option 1, for 12.4.1 at least!

The deprecated class is the thing that got us into this problem in the first place; option 2 will get users into weird places where things like is_a? don't work as expected, and don't emit a warning.

**Comment by jaym** [comment]
https://github.com/chef/chef/pull/3623
https://github.com/chef/chef/pull/3622

**Comment by jaym** [comment]
I'm also a fan of option 1 as it the most correct. Running both versions through CI

**Comment by lamont-granquist** [comment]
Ah yeah, I guess I'm in 1 then.  Breaking resource.is_a?(Thing) is going to break someone.

**Comment by thommay** [comment]
yeah, 1 sounds good.

**Comment by martinisoft** [comment]
Prefer option 1 as well. I've seen the resource.is_a? pattern in a few places.

**Comment by jaym** [comment]
closed through #3623
